File: 054/cpp/main.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        auto [left, top, right, bottom] = std::make_tuple(-1, -1, (int) matrix[0].size(), (int) matrix.size());
        int direction = 2;
        int i = 0, j = 0;
        vector<int> result;
        while (left + 1 < right && top + 1 < bottom) {
            result.push_back(matrix[i][j]);
            switch (direction) {
                case 0: j--; break;
                case 1: i--; break;
                case 2: j++; break;
                case 3: i++; break;
            }
            if (j == left) {
                bottom--;
                direction = 1;
                i--; j++;
            } else if (i == top) {
                left++;
                direction = 2;
                i++; j++;
            } else if (j == right) {
                top++;
                direction = 3;
                i++; j--;
            } else if (i == bottom) {
                right--;
                direction = 0;
                i--; j--;
            }
        }
        return result;
    }
};

File: 054/rust/main.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let (mut left, mut top, mut right, mut bottom) = (-1, -1, matrix[0].len() as i32, matrix.len() as i32);
        let mut direction = 2;
        let (mut i, mut j) = (0, 0);
        let mut result = Vec::new();
        while left + 1 < right && top + 1 < bottom {
            result.push(matrix[i as usize][j as usize]);
            match direction {
                0 => j -= 1,
                1 => i -= 1,
                2 => j += 1,
                3 => i += 1,
                _ => ()
            }
            if j == left {
                bottom -= 1;
                direction = 1;
                i -= 1;
                j += 1;
            } else if i == top {
                left += 1;
                direction = 2;
                i += 1;
                j += 1;
            } else if j == right {
                top += 1;
                direction = 3;
                i += 1;
                j -= 1;
            } else if i == bottom {
                right -= 1;
                direction = 0;
                i -= 1;
                j -= 1;
            }
        }
        return result;
    }
}
